### PR TITLE
Camo and Portrait Chooser

### PR DESCRIPTION
**Pull request by SJuliez**
Gives the image choosers an overhaul and their own package.
They...
- Show more than one column
- Show images of subdirectories (see screenshot)
- The camo chooser has a search bar that will search for any directory and/or filename match when at least 3 characters are entered
- Remember size, position and tree/image split
- Upon showing, will refresh the camo/portrait directory to update changed files

Also
- moves all camo/portrait creation to statics CamoManager and PortraitManager which handle image size and make sure they always return some Image, not null
- corrects portrait display to preserve the aspect ratio in the lobby, dialogs, UnitDisplay and tooltip

The AbstractImageChooser base class is not limited to showing images; it can be used for any tree selection. Maybe theres another application somewhere (RATs?)

Camos (all Kurita with subdirs)
image
Portraits
image
Searching "cava" - tooltip visible
image

### CODE REVIEW

**Comment by NickAragua** [comment]
Thie camo and portrait choice dialog is also used in MekHQ. Does it still work there after these changes are applied?

**Comment by Windchild292** [comment]
Different dialogs currently

**Comment by Windchild292** [comment]
Although I'm not sure with regards to resource file usage.

**Comment by SJuliez** [comment]
MekHQ has its own copy of CamoChoiceDialog (which isnt even exactly the same as the former one in MM). I replaced all the camos = new DirectoryItems(...) etc. from the MM classes with calls to the CamoManager but any class could still load the directory on its own. It's just not necessary. I'll admit though that I don't have MekHQ cloned.

**Comment by Windchild292** [comment]
I'll handle MekHQ's side, and once this is merged make the necessary adjustments to swap how we save our icons

File: megamek/src/megamek/client/ui/swing/ChatLounge.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/ChatLounge.java:180]: Missing access modifier

File: megamek/src/megamek/client/ui/swing/ChatLounge.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/ChatLounge.java:469]:
```suggestion
                if ((result == JOptionPane.CANCEL_OPTION) || (camoDialog.getSelectedItem() == null)) {
```

File: megamek/src/megamek/client/ui/swing/ChatLounge.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/ChatLounge.java:1945]: Remove commented code

File: megamek/src/megamek/client/ui/swing/ChatLounge.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/ChatLounge.java:2366]:
```suggestion
        if ((result == JOptionPane.CANCEL_OPTION) || (mcd.getSelectedItem() == null)) {
```

File: megamek/src/megamek/client/ui/swing/ChatLounge.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/ChatLounge.java:2399]:
```suggestion
                || (entity.getOwnerId() == clientgui.getClient().getLocalPlayer().getId());
```

Comment (Windchild292) [reply megamek/src/megamek/client/ui/swing/ChatLounge.java:2399]: Just a suggestion that I find makes it easier to read

File: megamek/src/megamek/client/ui/swing/ClientGUI.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/ClientGUI.java:1855]:
```suggestion
        if ((entity.getCamoCategory() != null) && !entity.getCamoCategory().equals(IPlayer.NO_CAMO)) {
```

File: megamek/src/megamek/client/ui/swing/ClientGUI.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/ClientGUI.java:1859]: ?

File: megamek/src/megamek/common/util/fileUtils/DirectoryItem.java
Comment (Windchild292) [inline megamek/src/megamek/common/util/fileUtils/DirectoryItem.java:36]:
```suggestion
            DirectoryItem dOther = (DirectoryItem) other;

```

File: megamek/src/megamek/common/util/fileUtils/DirectoryItem.java
Comment (Windchild292) [inline megamek/src/megamek/common/util/fileUtils/DirectoryItem.java:35]:
```suggestion
        } else if (other instanceof DirectoryItem) {

```

File: megamek/src/megamek/common/util/fileUtils/DirectoryItem.java
Comment (Windchild292) [inline megamek/src/megamek/common/util/fileUtils/DirectoryItem.java:39]:
```suggestion
        } else {
            return false;
        }
```

File: megamek/src/megamek/client/ui/swing/tileset/TilesetManager.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/tileset/TilesetManager.java:621]:
```suggestion
        if ((entity.getCamoCategory() != null) && !entity.getCamoCategory().equals(IPlayer.NO_CAMO)) {
```

File: megamek/src/megamek/client/ui/swing/skinEditor/SkinEditorMainGUI.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/skinEditor/SkinEditorMainGUI.java:875]:
```suggestion
        if ((entity.getCamoCategory() != null) && !entity.getCamoCategory().equals(IPlayer.NO_CAMO)) {
```

File: megamek/src/megamek/client/ui/swing/CustomPilotView.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/CustomPilotView.java:437]:
```suggestion
        } else {
            return portraitDialog.getSelectedItem().getItem();
        }
```

File: megamek/src/megamek/client/ui/swing/CustomPilotView.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/CustomPilotView.java:430]:
```suggestion
        } else {
            return portraitDialog.getSelectedItem().getCategory();
        }
```

File: megamek/src/megamek/client/ui/swing/CustomPilotView.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/CustomPilotView.java:111]:
```suggestion
                        ((JButton) e.getSource()).setIcon(PortraitManager.getPortraitIcon(category, filename));
```

File: megamek/src/megamek/client/ui/swing/ScenarioDialog.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/ScenarioDialog.java:92]:
```suggestion
                    if ((result == JOptionPane.CANCEL_OPTION) || (camoDialog.getSelectedItem() == null)) {
```

File: megamek/src/megamek/client/ui/swing/boardview/EntitySprite.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/boardview/EntitySprite.java:931]:
```suggestion
                        addToTT("PilotPortraitW", BR, imagePath, (int) h);
```

File: megamek/src/megamek/client/ui/swing/boardview/EntitySprite.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/boardview/EntitySprite.java:934]:
```suggestion
                        addToTT("PilotPortraitH", BR, imagePath, (int) w);
```

File: megamek/src/megamek/client/ui/swing/dialog/imageChooser/ImageList.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/dialog/imageChooser/ImageList.java:65]: Could this be a list instead of an ArrayList?

File: megamek/src/megamek/client/ui/swing/dialog/imageChooser/CamoRenderer.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/dialog/imageChooser/CamoRenderer.java:42]: Is this listed anywhere else, and do we support and file types other than these?

File: megamek/src/megamek/client/ui/swing/dialog/imageChooser/PortraitChooser.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/dialog/imageChooser/PortraitChooser.java:73]: Can this be a list?

File: megamek/src/megamek/client/ui/swing/tileset/CamoManager.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/tileset/CamoManager.java:52]: This looks odd

File: megamek/src/megamek/client/ui/swing/tileset/CamoManager.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/tileset/CamoManager.java:42]: Refactor to MegaMek.getLogger()

File: megamek/src/megamek/client/ui/swing/tileset/CamoManager.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/tileset/CamoManager.java:114]:
```suggestion
        if ((category == null) || (name == null)) {
```

File: megamek/src/megamek/client/ui/swing/tileset/CamoManager.java
Comment (Windchild292) [reply megamek/src/megamek/client/ui/swing/tileset/CamoManager.java:42]: Actually, this whole file needs logger standardization to MegaMek.getLogger()

File: megamek/src/megamek/client/ui/swing/tileset/PortraitManager.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/tileset/PortraitManager.java:42]: This whole file can be standardized to a MegaMek.getLogger()

File: megamek/src/megamek/client/ui/swing/tileset/PortraitManager.java
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/tileset/PortraitManager.java:120]:
```suggestion
        if ((category == null)
                || (name == null)
```

**Review by Windchild292 — CHANGES_REQUESTED** [review]
Some changes, mostly formatting, requested.

Otherwise, I'll port this to MekHQ and standardize to the new icon format in the near future once this is merged.

File: megamek/src/megamek/client/ui/swing/ClientGUI.java
Comment (SJuliez) [reply megamek/src/megamek/client/ui/swing/ClientGUI.java:1859]: The following code can probably be simplified. But this PR is already much bigger than I originally planned and I needed to stop at some place. This is the place.

File: megamek/src/megamek/client/ui/swing/CustomPilotView.java
Comment (SJuliez) [reply megamek/src/megamek/client/ui/swing/CustomPilotView.java:430]: Eh, this return is the "normal" return value, so I'd rather not put it in an else clause. Purely subjective.

File: megamek/src/megamek/client/ui/swing/CustomPilotView.java
Comment (SJuliez) [reply megamek/src/megamek/client/ui/swing/CustomPilotView.java:437]: Same as above

File: megamek/src/megamek/client/ui/swing/dialog/imageChooser/CamoRenderer.java
Comment (SJuliez) [reply megamek/src/megamek/client/ui/swing/dialog/imageChooser/CamoRenderer.java:42]: The ImageFileFactory and ScaledImageFileFactory support only these extensions and have their own Strings for them, but not as an array. The worst that can happen here is that the file extension is shown instead of hidden.

File: megamek/src/megamek/client/ui/swing/dialog/imageChooser/ImageList.java
Comment (SJuliez) [reply megamek/src/megamek/client/ui/swing/dialog/imageChooser/ImageList.java]: Yes, changed.

File: megamek/src/megamek/client/ui/swing/dialog/imageChooser/PortraitChooser.java
Comment (SJuliez) [reply megamek/src/megamek/client/ui/swing/dialog/imageChooser/PortraitChooser.java]: Yes, changed.

**Comment by SJuliez** [comment]
OK, added in almost all suggestions and corrections, commented the remaining 2 or so.

**Review by Windchild292 — APPROVED** [review]
Approved pending CI

**Pull request merged**
